chore: weaken commutativity assumptions for AdjoinRoot.lift and AdjoinRoot.liftHom

AntoineChambert-Loir opened the pull request:
I weaken a commutativity assumption for docs#AdjoinRoot.liftHom that the target algebra be commutative. It is only assumed to be a Semiring.
For that, I need to generalize docs#AdjoinRoot.lift which takes `i : R ->+* S`, `a : S` with `f.eval₂ i a = 0` and defines `AdjoinRoot f ->+* S` that extends `i ` and sends `AdjoinRoot.root f` to `a`.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->


The initial version of the PR was to define `AdjoinRoot.lift'` where `S` is only a Semiring, with the additional assumption `hcomm : ∀ r, Commute (i r) a` that the element `a` commutes with the image of `i`.

@jcommelin suggested to provide this argument as an `autoParam`, filled in by an ad hoc `commutativity` tactic. However, such an argument cannot be made implicit and a first version required to have this argument given each time, most of the time by `_`, even for all subsequent lemmas, `AdjoinRoot.lift_mk`, etc.

To avoid this, I tried to make this assumption a `⦃hcomm : ∀ r, Commute (i r) a⦄`. This appears to have some side effects that proofs using an `exact` tactic or in term mode do not work well, but one can resort on `by simp […]`. Maybe @eric-wieser or @Vierkantor will have opinions on that. In some cases one can still give the argument as `(hcomm := by commutativity)`.

(There are 3 options :
- separate lemmas for the fully commutative case and for the case where a proof is needed
- using the `commutativity` tactic, but with an argument `_` sometimes
- using the optional argument.)


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Commit:
remove commutativity assumptions for AdjoinRoot

Commit:
lint100

Commit:
cleanup docstring

Commit:
explicitly add the theoretically implicit argument

eric-wieser commented on Mathlib/RingTheory/AdjoinRoot.lean:
```diff
@@ -266,48 +272,57 @@ theorem of.injective_of_degree_ne_zero [IsDomain R] (hf : f.degree ≠ 0) :
-/-- Lift a ring homomorphism `i : R →+* S` to `AdjoinRoot f →+* S`. -/
-def lift (i : R →+* S) (x : S) (h : f.eval₂ i x = 0) : AdjoinRoot f →+* S := by
-  apply Ideal.Quotient.lift _ (eval₂RingHom i x)
+/-- Given `s : S`, lift a ring homomorphism `i : R →+* S` whose image commutes
+ with `s`
```
```suggestion
/-- Given `x : S`, lift a ring homomorphism `i : R →+* S` whose image commutes
 with `x`
```

Commit:
switch order of arguments, but do nothing (attempts failed)

AntoineChambert-Loir commented on Mathlib/FieldTheory/Adjoin.lean:
```diff
@@ -991,7 +991,8 @@ noncomputable def adjoinRootEquivAdjoin (h : IsIntegral F α) :
     adjoinRootEquivAdjoin F h (AdjoinRoot.root (minpoly F α)) = AdjoinSimple.gen F α :=
-  AdjoinRoot.lift_root (aeval_gen_minpoly F α)
+   -- It is a bit annoying that one be obliged to fill in this implicit hypothesis
+  AdjoinRoot.lift_root (hc := fun _ ↦ Commute.all _ _) (aeval_gen_minpoly F α)
```
@jcommelin I tried to think about this, but got confused by the syntax and could not make any progress (what I tried did not work at all). From the Zulip discussion, it seems that you had an idea of how that could be implemented… Could you please share some hint?

jcommelin replied:
I pushed to `jmc-AdjoinRoot`. I'll let you decide which option you prefer.

(Note that I renamed `lift'` to `lift`, and there is now only one version.)

AntoineChambert-Loir replied:
I just had a look! Wonderful!

Commit:
make the commutativity argument implicit whenever possible (proofs are ugly)

Commit:
adjust docstring and FieldTheory/Adjoin.lean

Commit:
linter 100

AntoineChambert-Loir commented:
The `Adjoin.lift` function now allows that target to be noncommutative, it only needs a proof that the element that `Adjoin.root f` maps to commutes with the image of the ring. 
Thanks to a tactic `commutativity` provided by @jcommelin , this argument can be filled automatically.
In all subsequent rewriting lemmas, I made it ⦃hcomm : ∀ r, Commute (i r) a⦄, but the defect is that I couldn't really use `exact` tactics, or term-mode proofs, so that the proofs are now a bit ugly, using a lot of `by simp […]`.

Commit:
adjust docstring in `AdjoinRoot.lift`

Commit:
Trying to make simpNF linter happy

Vierkantor committed:
Use `aesop` to search for commutativity proofs

This allows us to restore quite a few uses of `lift` to the original. The only drawback in the commutative case is we have to write `(lift i a h) x` instead of `lift i a h x` (since the autoparam comes after the `h`). Also, `simp` doesn't discharge autoparams automatically, so I added an extra set of lemmas that use implicits instead of autoparams, to be filled by unification.

Vierkantor commented:
I pushed a commit which tags a few lemmas and replaces `commutativity` with a call to `aesop`. It's almost as good as original for the commutative case, if you ignore the parentheses :)

eric-wieser commented on Mathlib/RingTheory/AdjoinRoot.lean:
```diff
@@ -266,43 +269,81 @@ theorem of.injective_of_degree_ne_zero [IsDomain R] (hf : f.degree ≠ 0) :
-  apply Ideal.Quotient.lift _ (eval₂RingHom i x)
+It is currently only used in `AdjoinRoot.lift`. -/
+macro "commutativity" : tactic =>
+  `(tactic| first | aesop |
```
This should probably raise an error if there is no `Commute` in the goal

jcommelin replied:
@eric-wieser Do you know how to check this? I don't...

eric-wieser replied:
Are Aesop rulesets a thing? If not, using `guard_target` would probably work.

Vierkantor replied:
`guard_target` isn't quite what we want here: for example, `lift` wants to use `commutativity` to solve `∀ r, Commute (i r) x`.

WRT rulesets: do you mean we should set the `commutativity` tactic to use a commutativity ruleset exclusively? I feel like that's premature optimization.

eric-wieser replied:
Yes, I think we should do that because it matches the pattern set by `measurability`. I don't think that `commutativity` as a tactic should be solving everything that `aesop` can solve.

AntoineChambert-Loir replied:
@eric-wieser @Vierkantor do you want me to try copying the measurability tactic file and build a commutativity tactic along its line?

AntoineChambert-Loir committed:
Merge branch 'master' into ACL/AdjoinRoot

AntoineChambert-Loir committed:
try to add commutativity tactic

AntoineChambert-Loir committed:
import commutativity

AntoineChambert-Loir committed:
update Mathlib.lean

AntoineChambert-Loir committed:
make linter 100 happy

AntoineChambert-Loir committed:
update Tactic.lean

AntoineChambert-Loir committed:
try

AntoineChambert-Loir committed:
try to extend commutativity

Vierkantor commented on Mathlib/Tactic/Commutativity.lean:
```diff
@@ -0,0 +1,46 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Lori (based on Measurability, by Miyahara Kō)
```
Typo? :)
```suggestion
Authors: Antoine Chambert-Loir (based on Measurability, by Miyahara Kō)
```
